### PR TITLE
RCNN w/ FPN, Sync BN & Disable static alloc for RCNN

### PR DESCRIPTION
**Pull request by Jerryzcn**
* Faster RCNN w/ FPN (refactored from @Angzz ) model zoo not uploaded yet
  * e2e_fasterrcnn_fpn_resnet50_v1b_2x_coco17 mAP: **0.384** | caffe baseline: 0.379
  * e2e_fasterrcnn_fpn_bn_resnet50_v1b_2x_coco17 mAP: **0.393** | caffe baseline: N/A
  * e2e_fasterrcnn_fpn_resnet101_v1d_2x_coco17 mAP: **0.412** caffe baseline: 0.398
* Mask RCNN w/ FPN model zoo not uploaded yet
  * e2e_maskrcnn_fpn_resnet50_v1b_2x_coco17 boxAP: 0.392, segmAP: **0.353** | caffe boxAP baseline: 0.386 caffe segmAP baseline: 0.345
  * e2e_maskrcnn_fpn_resnet101_v1d_2x_coco17 boxAP: 0.423, segmAP: **0.377** | caffe boxAP baseline: 0.409 caffe segmAP baseline: 0.364
* disabled static alloc for RCNN to save memory 
* added mask rcnn resnet 101 v1b

### CODE REVIEW

File: gluoncv/data/transforms/presets/rcnn.py [hunk 1]
```diff
@@ -130,7 +130,7 @@ class FasterRCNNDefaultTrainTransform(object):
     def __init__(self, short=600, max_size=1000, net=None, mean=(0.485, 0.456, 0.406),
                  std=(0.229, 0.224, 0.225), box_norm=(1., 1., 1., 1.),
                  num_sample=256, pos_iou_thresh=0.7, neg_iou_thresh=0.3,
-                 pos_ratio=0.5, **kwargs):
+                 pos_ratio=0.5, ashape=128, **kwargs):
```
Comment (zhreshold) [inline gluoncv/data/transforms/presets/rcnn.py]: if move to args, need doc

File: gluoncv/model_zoo/faster_rcnn/faster_rcnn.py [hunk 2]
```diff
@@ -418,9 +431,72 @@ def faster_rcnn_resnet50_v1b_coco(pretrained=False, pretrained_base=True, **kwar
...
+    pretrained_base = False if pretrained else pretrained_base
+    norm_layer = mx.gluon.contrib.nn.SyncBatchNorm if syncbn else nn.BatchNorm
+    base_network = resnet50_v1b(pretrained=pretrained_base, dilated=True,
+                                use_global_stats=not syncbn, norm_layer=norm_layer)
+    features = nn.HybridSequential()
+    top_features = nn.HybridSequential()
+    box_features = nn.HybridSequential()
+    for layer in ['conv1', 'bn1', 'relu', 'maxpool', 'layer1', 'layer2', 'layer3', 'layer4']:
+        features.add(getattr(base_network, layer))
+        box_features.add(nn.Conv2D(256, 1, weight_initializer=mx.init.Xavier(factor_type='in'),
```
Comment (zhreshold) [inline gluoncv/model_zoo/faster_rcnn/faster_rcnn.py]: box_features adding layers in for loop?

File: gluoncv/model_zoo/rcnn/roi_heads.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+"""ROI heads."""
+
+from __future__ import absolute_import
+
+from mxnet.gluon import nn
+
+
+class MLPFeatureExtractor(nn.HybridBlock):
+    """
+    Head for FPN for classification and box regression.
```
Comment (zhreshold) [inline gluoncv/model_zoo/rcnn/roi_heads.py]: missing doc for arguments

File: gluoncv/model_zoo/faster_rcnn/faster_rcnn.py [hunk 4]
```diff
@@ -142,12 +146,13 @@ def __init__(self, features, top_features, classes,
                  rpn_train_pre_nms=12000, rpn_train_post_nms=2000,
                  rpn_test_pre_nms=6000, rpn_test_post_nms=300, rpn_min_size=16,
                  num_sample=128, pos_iou_thresh=0.5, pos_ratio=0.25, max_num_gt=300,
-                 additional_output=False, **kwargs):
+                 additional_output=False, syncbn=False, **kwargs):
```
Comment (zhreshold) [inline gluoncv/model_zoo/faster_rcnn/faster_rcnn.py]: need doc for syncbn

File: gluoncv/model_zoo/model_zoo.py [hunk 5]
```diff
@@ -15,6 +15,8 @@
 from .resnext import *
 from .senet import *
 from .se_resnet import *
+from .superresnet import *
```
Comment (zhreshold) [inline gluoncv/model_zoo/model_zoo.py]: where is superresnet?

File: gluoncv/model_zoo/faster_rcnn/faster_rcnn.py (hunk 2 above)
Comment (Jerryzcn) [reply gluoncv/model_zoo/faster_rcnn/faster_rcnn.py]: this is an error

File: gluoncv/model_zoo/model_zoo.py (hunk 5 above)
Comment (Jerryzcn) [reply gluoncv/model_zoo/model_zoo.py]: local experiment will remove

File: gluoncv/model_zoo/rcnn/roi_heads.py (hunk 3 above)
Comment (Jerryzcn) [reply gluoncv/model_zoo/rcnn/roi_heads.py]: okay.. this is not supposed to be here...

**Comment by mli** [comment]
Job PR-494-5 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-494/5/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-494-6 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-494/6/index.html 
Code coverage of this PR:  vs. Master:

**Comment by YutingZhang** [comment]
Awesome! FPN is here for Faster RCNN.

**Comment by mli** [comment]
Job PR-494-28 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-494/28/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-494-29 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-494/29/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-494-30 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-494/30/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-494-31 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-494/31/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-494-33 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-494/33/index.html 
Code coverage of this PR:  vs. Master:

**Comment by zhreshold** [comment]
@Jerryzcn Is the PR ready for review/merge?

**Comment by Jerryzcn** [comment]
Yes, it is ready.

File: gluoncv/model_zoo/faster_rcnn/faster_rcnn.py [hunk 6]
```diff
@@ -181,6 +204,62 @@ def reset_class(self, classes):
         super(FasterRCNN, self).reset_class(classes)
         self._target_generator = {RCNNTargetGenerator(self.num_class)}
 
+    def _pyramid_roi_feats(self, F, features, rpn_rois, roi_size, strides, roi_mode='pool',
```
Comment (zhreshold) [inline gluoncv/model_zoo/faster_rcnn/faster_rcnn.py]: default roi_mode is 'align' right?

**Comment by zhreshold** [comment]
The rest LGTM! Is the pre-trained weights ready  on s3 for test?

File: gluoncv/model_zoo/faster_rcnn/faster_rcnn.py (hunk 6 above)
Comment (Jerryzcn) [reply gluoncv/model_zoo/faster_rcnn/faster_rcnn.py]: good point

**Comment by Jerryzcn** [comment]
not yet

**Comment by mli** [comment]
Job PR-494-37 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-494/37/index.html 
Code coverage of this PR:  vs. Master:

**Comment by zhreshold** [comment]
LGTM, I can merge it now and once the weights are available please let me know and submit a new PR with hashtags.

**Comment by mli** [comment]
Job PR-494-38 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-494/38/index.html 
Code coverage of this PR:  vs. Master:

**Comment by Jerryzcn** [comment]
I just notice there is a bug in FPNFeatureExpander, the final anti-aliasing conv is not used in the outputs

**Comment by YutingZhang** [comment]
@Jerryzcn Is this ready to merge?

**Comment by mli** [comment]
Job PR-494-50 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-494/50/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-494-53 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-494/53/index.html 
Code coverage of this PR:  vs. Master:

File: gluoncv/model_zoo/rpn/rpn.py [hunk 7]
```diff
@@ -52,54 +53,175 @@ class RPN(gluon.HybridBlock):
...
-            self.conv1 = nn.HybridSequential()
-            self.conv1.add(nn.Conv2D(channels, 3, 1, 1, weight_initializer=weight_initializer))
-            self.conv1.add(nn.Activation('relu'))
-            # use sigmoid instead of softmax, reduce channel numbers
-            self.score = nn.Conv2D(anchor_depth, 1, 1, 0, weight_initializer=weight_initializer)
-            self.loc = nn.Conv2D(anchor_depth * 4, 1, 1, 0, weight_initializer=weight_initializer)
+            if self._multi_level:
+                asz = alloc_size
+                self.anchor_generator = nn.HybridSequential()
+                for _, st, s in zip(range(num_stages), strides, scales):
```
Comment (HannibalAPE) [inline gluoncv/model_zoo/rpn/rpn.py]: hey, I think there may be a little mistake?
every stage_anchor_generator should be initialized with scales=(2, 4, 8, 16, 32), not just one of them.
I ran your codes to train mask_rcnn with fpn, but got very huge loss.
After I changed the scales problem, the loss looks common now.

by the way, is the pretrained model of mask_rcnn_fpn_resnet50_v1b_coco available now?
 I got a mistake when loading the params you provided, which is 
'Parameter 'box_featueres.1.weight' is missing in file './models/mask_rcnn_fpn_resnet50_v1b_coco-1364d0af.params', which contains parameters: 'features.resnetv1b_conv0_weight', 'features.resnetv1b_batchnorm0_gamma', 'features.resnetv1b_batchnorm0_beta', ..., 'mask.deconv.8.weight', 'mask.deconv.8.bias', 'mask.mask.weight', 'mask.mask.bias'. Set allow_missing=True to ignore missing parameters.'
